recipes-bsp/secure-boot-otp-provisioning/README.md

```diff
@@ -0,0 +1,149 @@
+```
+
+Reboot the IOT2050 and follow the instruction on the device. The debug serial
+port has to be connected to give your consent to the provisioning routine.
+
+> Warning: after provisioning, all application must be signed with the currently
+> using key.
+
+The key provisioning just need be run once. Later secure boot firmware building
+should drop the `kas/opt/key-provision.yml`
+
+## Revoke the leaked currently using key
+
+If unfortunately the currently using key is leaked, you could switch to use the
+next keysets as the effective one.
+
+First make sure the new key is copied to `recipes-bsp/u-boot/files/keys/custMpk.pem`.
+
+Then start to building the new key signed firmware together with the key switching
+otpcmd data.
+
+```shell
+./kas-container build kas-iot2050-boot-pg2.yml:kas/opt/secure-boot.yml:kas/opt/key-switch.yml
+```
+
+> Warning: The default switching within kas/opt/key-switch.yml is from MPK to SMPK.
+> If you want to switch from SMPK to BMPK, please refer to
+[Flexible key provisioning](#flexible-key-provisioning).
+
+Then follow the exact same steps as defined in above, to update the firmware and
+follow the instructions on device after rebooting.
+
+## Flexible key provisioning
+
+The otpcmd data produced by the default `kas/opt/key-provision.yml` contains
+
+- the MPK and SMPK key sets
+- control bit that enabling the secure boot
+
+Sometimes it would be helpful to have different configurations. This could be done
+by passing different parameters to:
+`recipes-bsp/secure-boot-otp-provisioning/files/make-otpcmd.sh`
+and have different configurations in
+`recipes-bsp/secure-boot-otp-provisioning/files/its/key-provision.its`
+
+### Provision the additional BMPK
+
+If the BMPK need to be programmed together, before building the firmware, link
+the `key-provision.its` to `key-provision-3keys.its.template`:
+
+```bash
+# cd to files/its/
+ln -sf key-provision-3keys.its.template key-provision.its
+```
+
+then uncomment below line in `meta-iot2050/kas/opt/key-provision.yml`:
+
+```
+# OTP_BMPK ?= "./keys/custBmpk.pem"
+```
+
+### Provision keys only without enabling secure boot
+
+If only the keys need to be programmed but not enabling the secure boot, before
+building the firmware, link the
+`key-provision.its` to `key-provision-only-keys.its.template`.
+
+```bash
+# cd to files/its/
+ln -sf key-provision-only-keys.its.template key-provision.its
+```
+
+### Enable secure boot only
+
+If you want to only enable the secure boot, (because the keys are already programmed)
+just before building the firmware, link the
+`key-provision.its` to `key-provision-only-enable.its.template`.
+
+```bash
+# cd to files/its/
+ln -sf key-provision-only-enable.its.template key-provision.its
+```
+
+### Switch from SMPK to BMPK
+
+If the key switching is from SMPK to BMPK, uncomment below lines in
+`meta-iot2050/kas/opt/key-switch.yml`:
+
+```
+# OTP_MPK = ""
+# OTP_SMPK = "./keys/custSmpk.pem"
+# OTP_BMPK = "./keys/custBmpk.pem"
+```
```

Comment:
Can't we have a script that - independently of the image build - generates the required otpcmd block, signs it, and then either allows to transfer that to running devices or to inject it - post build - into a firmware image file for first-boot provisioning? The less we model these sensitive steps inside the build, the better.

Reply:
That's exactly the `recipes-bsp/secure-boot-otp-provisioning/files/make-otpcmd.sh` is doing now.

When talking `build`, what exactly are you referring to? meta-iot2050 or u-boot?

Reply:
Anything that happens during `kas-container build ...`. We should rather provide some means to inject a prebuilt & signed block into an image build - or do that post-build as I wrote elsewhere.

Reply:
Are you suggesting we should not add this script into meta-iot2050? but provide it as a standalone script together with the usage document?